zebra/zebra_rib.c

```diff
@@ -2113,6 +2358,179 @@ static void rib_process_result(struct zebra_dplane_ctx *ctx)
 	dplane_ctx_fini(&ctx);
 }
 
+/*
+ * Handle notification from async dataplane: the dataplane has detected
+ * some change to a route, and notifies zebra so that the control plane
+ * can reflect that change.
+ */
+static void rib_process_dplane_notify(struct zebra_dplane_ctx *ctx)
+{
+	struct route_node *rn = NULL;
+	struct route_entry *re = NULL;
+	struct nexthop *nexthop;
+	char dest_str[PREFIX_STRLEN] = "";
+	const struct prefix *dest_pfx, *src_pfx;
+	rib_dest_t *dest;
+	bool fib_changed = false;
+	bool debug_p = IS_ZEBRA_DEBUG_DPLANE | IS_ZEBRA_DEBUG_RIB;
+	int start_count, end_count;
+	dest_pfx = dplane_ctx_get_dest(ctx);
+
+	/* Note well: only capturing the prefix string if debug is enabled here;
+	 * unconditional log messages will have to generate the string.
+	 */
+	if (debug_p)
+		prefix2str(dest_pfx, dest_str, sizeof(dest_str));
+
+	/* Locate rn and re(s) from ctx */
+	rn = rib_find_rn_from_ctx(ctx);
+	if (rn == NULL) {
+		if (debug_p) {
+			zlog_debug("Failed to process dplane notification: no routes for %u:%s",
+				   dplane_ctx_get_vrf(ctx), dest_str);
+		}
+		goto done;
+	}
+
+	dest = rib_dest_from_rnode(rn);
+	srcdest_rnode_prefixes(rn, &dest_pfx, &src_pfx);
+
+	if (debug_p)
+		zlog_debug("%u:%s Processing dplane notif ctx %p",
+			   dplane_ctx_get_vrf(ctx), dest_str, ctx);
+
+	/*
+	 * Take a pass through the routes, look for matches with the context
+	 * info.
+	 */
+	RNODE_FOREACH_RE(rn, re) {
+		if (rib_route_match_ctx(re, ctx, false /*!update*/))
+			break;
+	}
+
+	/* No match? Nothing we can do */
+	if (re == NULL) {
+		if (debug_p)
+			zlog_debug("%u:%s Unable to process dplane notification: no entry for type %s",
+				   dplane_ctx_get_vrf(ctx), dest_str,
+				   zebra_route_string(
+					   dplane_ctx_get_type(ctx)));
+
+		goto done;
+	}
+
+	/* Is this a notification that ... matters? We only really care about
+	 * the route that is currently selected for installation.
+	 */
+	if (re != dest->selected_fib) {
```

Comment:
Why would we not care about un-selected route changes?
What if we try to select that route again?

Reply:
> Why would we not care about un-selected route changes?
> What if we try to select that route again?

so the thinking here is:
a) zebra is only interacting with the dataplane about the 'selected' route
b) anything that arrives from the dataplane about some other route isn't really the dataplane's business - it's zebra who drives the rib-to-fib decision-making
c) so any fib/dplane notification that arrives about other routes is stale in some way